### PR TITLE
The many flavours of WMI post

### PR DESCRIPTION
**Pull request by FranciscoNabas**
Sorry for the previous PR. Forgot to push the commit 🤦‍♂️

# PR Summary

post_title: The many flavours of WMI management
username: FranciscoNabas
categories: PowerShell
tags: PowerShell, WMI, .NET, COM, Windows Management
summary: Working with the different ways of managing WMI.

## PR Checklist

- Do you want this post to be published on a specific date?: Y/N [Specify date]:
- [X] I have read the [contributors guide][contrib] and followed the style and process guidelines
- [X] PR has a meaningful title
- [ ] Includes content related an open issue
  - see [Closing issues using keywords][key].
- [X] This PR is ready to merge and is not **Work in Progress**

[contrib]: https://docs.microsoft.com/powershell/scripting/community/contributing/overview
[key]: https://help.github.com/en/articles/closing-issues-using-keywords

### CODE REVIEW

File: Posts/2022/07/ReadCMStatusMessages.md [hunk 1]
```diff
@@ -1,375 +0,0 @@
----
-post_title: Reading Configuration Manager Status Messages With PowerShell
-username: francisconabas
-categories: PowerShell
-tags: SCCM, MECM, Status Message, Config Manager
-summary: This post's intent is to show how to read Configuration Manager status messages using WMI and Win32 API function FormatMessage.
----
```
Comment (sdwheeler) [inline Posts/2022/07/ReadCMStatusMessages.md:7]: This PR is deleting your previous post. Please undo the delete. This previous post should not be part of this PR.

File: Posts/2022/08/many-wmi-flavours.md [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+
+The _auto-complete_ feature in Visual Studio Code.
+
+![Auto-Complete with CIM](Get-CimInstance_autoComplete.png)
+
+### Calling a WMI Method
+
+The CIM Cmdlets introduces a unique way of calling WMI Methods.
+The results of a CIM query are called _CimInstance(s)_, and you cannot call instance methods like you would with the other two options.
+Instead, you call another Cmdlet called _Invoke-CimMethod_.
```
Comment (michaeltlombardi) [inline Posts/2022/08/many-wmi-flavours.md:255]:
```suggestion
The results of a CIM query are called **CimInstances**, and you cannot call instance methods like you would with the other two options.
Instead, you call another Cmdlet called `Invoke-CimMethod`.
```

File: Posts/2022/08/many-wmi-flavours.md [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+
+Path          : \\.\root\ScriptingBlogCoolNamespace:CustomClass.PropertyKey=1
+RelativePath  : CustomClass.PropertyKey=1
+Server        : .
+NamespacePath : root\ScriptingBlogCoolNamespace
+ClassName     : CustomClass
+IsClass       : False
+IsInstance    : True
+IsSingleton   : False
+```
```
Comment (michaeltlombardi) [inline Posts/2022/08/many-wmi-flavours.md:365]:
````suggestion
```powershell
$namespace = ([wmiclass]'root:__Namespace').CreateInstance()
$namespace.Name = 'ScriptingBlogCoolNamespace'
$namespace.Put()
```

```Output
Path          : \\.\root:__NAMESPACE.Name="ScriptingBlogCoolNamespace"
RelativePath  : __NAMESPACE.Name="ScriptingBlogCoolNamespace"
Server        : .
NamespacePath : root
ClassName     : __NAMESPACE
IsClass       : False
IsInstance    : True
IsSingleton   : False
```

```powershell
$class = [wmiclass]::new('root\ScriptingBlogCoolNamespace', '', $null)
$class['__Class'] = 'CustomClass'
$class.Qualifiers.Add('Static', $true)
$class.Properties.Add('Source', 'Custom class with .NET!')
$class.Properties.Add('PropertyKey', 1)
## You need a Key property, otherwise WMI wouldn't be able to assemble the path of a new instance.
$class.Properties['PropertyKey'].Qualifiers.Add('Key', $true)
$class.Put()
```

```Output
Path          : \\.\root\ScriptingBlogCoolNamespace:CustomClass
RelativePath  : CustomClass
Server        : .
NamespacePath : root\ScriptingBlogCoolNamespace
ClassName     : CustomClass
IsClass       : True
IsInstance    : False
IsSingleton   : False
```

```powershell
$instance = ([wmiclass]'root\ScriptingBlogCoolNamespace:CustomClass').CreateInstance()
$instance.Source = 'CustomInstance!'
$instance.Put()
```

```Output
Path          : \\.\root\ScriptingBlogCoolNamespace:CustomClass.PropertyKey=1
RelativePath  : CustomClass.PropertyKey=1
Server        : .
NamespacePath : root\ScriptingBlogCoolNamespace
ClassName     : CustomClass
IsClass       : False
IsInstance    : True
IsSingleton   : False
```
````

File: Posts/2022/08/many-wmi-flavours.md [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+
+And just like that, we have our own Namespace, Class and Instance!  
+The results on _WmiExplorer_.
```
Comment (michaeltlombardi) [inline Posts/2022/08/many-wmi-flavours.md:368]:
```suggestion
The results on **WmiExplorer**.
```

**Review by michaeltlombardi — CHANGES_REQUESTED** [review]
Thanks so much for this submission, @FranciscoNabas! This is a great post. My review mostly just makes minor corrections and standardizes the formatting.

File: Posts/2022/08/many-wmi-flavours.md [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+  
+Thank you for following along on this journey, and I see you next time!
+
+## Useful links
+
+[System.Management](https://docs.microsoft.com/en-us/dotnet/api/system.management?view=dotnet-plat-ext-6.0)  
+[WMI Scripting API](https://docs.microsoft.com/en-us/windows/win32/wmisdk/scripting-api-for-wmi)  
+[CIM Cmdlets](https://docs.microsoft.com/en-us/powershell/module/cimcmdlets/?view=powershell-7.2)  
+[Runtime-Callable Wrapper](https://docs.microsoft.com/en-us/dotnet/standard/native-interop/runtime-callable-wrapper)  
+[WMI Documentation](https://docs.microsoft.com/en-us/windows/win32/wmisdk/wmi-start-page)
```
Comment (michaeltlombardi) [inline Posts/2022/08/many-wmi-flavours.md:404]:
```suggestion
- [System.Management](https://docs.microsoft.com/en-us/dotnet/api/system.management?view=dotnet-plat-ext-6.0)  
- [WMI Scripting API](https://docs.microsoft.com/en-us/windows/win32/wmisdk/scripting-api-for-wmi)  
- [CIM Cmdlets](https://docs.microsoft.com/en-us/powershell/module/cimcmdlets/?view=powershell-7.2)  
- [Runtime-Callable Wrapper](https://docs.microsoft.com/en-us/dotnet/standard/native-interop/runtime-callable-wrapper)  
- [WMI Documentation](https://docs.microsoft.com/en-us/windows/win32/wmisdk/wmi-start-page)  
```

File: Posts/2022/08/many-wmi-flavours.md [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+Set-CimInstance -Namespace 'root\SMS\site_PS1' -Query "Select * From SMS_Collection Where Name = 'AwesomeDeviceCollection'" -Property @{ Name = 'AwesomeDeviceCollection_NewName' }
+
+#Or
+
+$collection = Get-CimInstance -Namespace 'root\SMS\site_PS1' -Query "Select * From SMS_Collection Where Name = 'AwesomeDeviceCollection'"
+$collection | Set-CimInstance -Property @{ Name = 'AwesomeDeviceCollection_NewName' }
+
+#Deleting
+
+$collection = Get-CimInstance -Namespace 'root\SMS\site_PS1' -Query "Select * From SMS_Collection Where Name = 'AwesomeDeviceCollection'"
```
Comment (michaeltlombardi) [inline Posts/2022/08/many-wmi-flavours.md:308]:
```suggestion
$params = @{
  Namespace = 'root\SMS\site_PS1'
  Query = "Select * From SMS_Collection Where Name = 'AwesomeDeviceCollection'"
}
$collection = Get-CimInstance @params
```

File: Posts/2022/08/many-wmi-flavours.md [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+
+Updating and deleting.
+
+```powershell
+Set-CimInstance -Namespace 'root\SMS\site_PS1' -Query "Select * From SMS_Collection Where Name = 'AwesomeDeviceCollection'" -Property @{ Name = 'AwesomeDeviceCollection_NewName' }
+
+#Or
+
+$collection = Get-CimInstance -Namespace 'root\SMS\site_PS1' -Query "Select * From SMS_Collection Where Name = 'AwesomeDeviceCollection'"
+$collection | Set-CimInstance -Property @{ Name = 'AwesomeDeviceCollection_NewName' }
```
Comment (michaeltlombardi) [inline Posts/2022/08/many-wmi-flavours.md:304]:
```suggestion
$params = @{
  Namespace = 'root\SMS\site_PS1'
  Query = "Select * From SMS_Collection Where Name = 'AwesomeDeviceCollection'"
}
$collection = Get-CimInstance @params
$collection | Set-CimInstance -Property @{
  Name = 'AwesomeDeviceCollection_NewName'
}
```

File: Posts/2022/08/many-wmi-flavours.md [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+Have a hard time remembering parameters? Me too! Luckily _auto-complete_ also works with them.
+
+![Auto-Complete with method parameters](Invoke-CimMethod_autoComplete.png)
+
+### Creating, Updating and Deleting a WMI Class Instance
+
+If you used the old WMI Cmdlets before, this will look familiar.
+
+```powershell
+$collection = New-CimInstance -Namespace 'root\SMS\site_PS1' -ClassName 'SMS_Collection' -Property @{ Name = 'AwesomeDeviceCollection'; LimitingCollectionID = 'PS1000042' }
```
Comment (michaeltlombardi) [inline Posts/2022/08/many-wmi-flavours.md:293]:
```suggestion
$params = @{
  Namespace = 'root\SMS\site_PS1'
  ClassName = 'SMS_Collection'
  Property = @{
    Name = 'AwesomeDeviceCollection'
    LimitingCollectionID = 'PS1000042'
  }
}
$collection = New-CimInstance @params
```

File: Posts/2022/08/many-wmi-flavours.md [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+Instead, you call another Cmdlet called `Invoke-CimMethod`.
+
+```powershell
+$commandLine = 'powershell.exe -ExecutionPolicy Bypass -Command "Write-Output ''Howdy! From WMI!''; Read-Host"'
+$result = Get-CimClass -ClassName 'Win32_Process'
+$output = $result | Invoke-CimMethod -MethodName 'Create' -Arguments @{ CommandLine = $commandLine }
+
+# Or
+
+$output = Invoke-CimMethod -ClassName 'Win32_Process' -MethodName 'Create' -Arguments @{ CommandLine = $commandLine }
```
Comment (michaeltlombardi) [inline Posts/2022/08/many-wmi-flavours.md:269]:
```suggestion
$params = @{
  MethodName = 'Create'
  Arguments = @{
    CommandLine = $commandLine
  }
}
$output = $result | Invoke-CimMethod @params

# Or
$params = @{
  ClassName = 'Win32_Process'
  MethodName = 'Create'
  Arguments = @{
    CommandLine = $commandLine
  }
}
$output = Invoke-CimMethod @params
```

**Review by michaeltlombardi — APPROVED** [review]
Minor cleanup for line length and formatting I missed on the first pass.

**Comment by sdwheeler** [comment]
Closing this PR. Posting submitted in #104
